rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,85 @@
+head and comparing it to a known-good file serialized at the previous release.
+This approach operates on the assumption that at HEAD we are able to produce
+a byte-identical serialized artifact, excluding custom header info and debug
+locations.
+
+Detecting forward incompatiblities can be accomplished by adding the following
+`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
+test files.
+
+```bash
+diff %s.bc <(stablehlo-opt %s | stablehlo-translate --serialize --target=0.X.0)
+```
+
+Passing through `stablehlo-opt` in this example removes debug info. An option
+to `stablehlo-translate` can be added to strip debug info and header info. This
+test should be run in CI on all PRs to prevent merging incompatible changes in
+StableHLO (_R1_) and LLVM revision bumps (_R2_).
+
+The StableHLO process for [Contributing Incompatible Changes](https://github.com/openxla/stablehlo/blob/main/docs/vhlo.md#add-versioned-serialization-test)
+requires a versioned test be serialized for each compatibility breaking change,
+which means that we can test forward compatibility between HEAD and each minor
+release of StableHLO. Given that all changes to the StableHLO opset, as well as
+changes to serialization machinery like targeting a newer version of the MLIR
+bytecode format, require bumping the minor version, there is a low chance of
+causing forward incompatibilities between different patch versions of StableHLO
+(i.e. `0.10.1 → 0.10.2`). Ensuring compatibility between HEAD and all minor
+releases (`0.9.0`, `0.10.0`, etc.) should provide thorough enough coverage of the
+two mentioned sources of forward incompatibilities.
+
+Other benefits of this approach include that it is lightweight (_R3_), can be run
+locally to detect compatibility issues during development, and it fits nicely into
+existing CI build-and-test jobs. If the underlying byte-identical assumption is
+invalidated, we can fall back to the alternate design in a hybrid testing approach,
+since byte-comparison will safely identify all forward incompatibilities, with a
+slight risk of false positives that should be tested using the alternate design.
+
+## Alternate Design: Serialize at HEAD, Deserialize at Target Release
+
+This design is explored more in a [previous commit of this RFC](https://github.com/openxla/stablehlo/blob/0792eb75e85c54f9d106878569b088d03c568b70/rfcs/20230517-forward-compatibility-testing.md#preferred-design-serialize-at-head-deserialize-at-target-release),
+and is summarized in this version for brevity. If byte-equality cannot be safely
+relied on, then we may need to use this approach, or a hybrid of the two. This
+design can be summarized as:
+
+* Maintain bytecode files serialized at HEAD, which are re-serialized before
+  each StableHLO release (~2x/wk): `stablehlo_legalize_to_vhlo.0_X_0.mlirbc.HEAD`
```

Comment:
Fwiw the related files in the GitHub repository are currently called `stablehlo_legalize_to_vhlo.0_X_0.mlir.bc`. (With a period between "mlir" and "bc").

Reply:
Good catch, at the same time I was working on a PR to rename to `.mlirbc` since the MLIR vscode extension recognizes this, but I don't think there is much benefit to being able to look at/edit the textual assembly of VHLO so dropped it